### PR TITLE
conformance: Fix logic in HTTPRouteResponseHeaderModifier test

### LINKED ISSUE #1384: Improve `ExpectedResponse` used in conformance tests
**What would you like to be added**:
A new type `Response` should be added containing the following fields (non-exhaustive):
```golang
// This struct represents a HTTPResponse returned from a backend.
type Response struct {
	StatusCode    int
	Headers       map[string]string
	AbsentHeaders []string
}
```
The `ExpectedResponse` struct should be modified to have field of the type `Response` replacing the current field `StatusCode`.

**Why this is needed**:
The `ExpectedResponse` contains a lot of information other than about _just_ the expected HTTP response, like the expected modified request, etc. It'd be nice to have a clear level of separation between the info specific to the other functionality of Gateway API and info related to the actual HTTP response.

### LINKED ISSUE #1473: HTTPRouteResponseHeaderModifier test seems incorrect
**What happened**:

I tried to run the latest conformance tests

**What you expected to happen**:

It failed, but I don't think it should have.

**How to reproduce it (as minimally and precisely as possible)**:

You can run the tests with https://github.com/istio/istio/pull/41353 I suppose

**Anything else we need to know?**:

Aside from trivial issue in https://github.com/kubernetes-sigs/gateway-api/pull/1472, the assertions seem generally wrong?

For example:

```yaml
  - matches:
    - path:
        type: PathPrefix
        value: /set
    filters:
    - type: ResponseHeaderModifier
      responseHeaderModifier:
        set:
        - name: X-Header-Set
          value: set-overwrites-values
```

```go
{
  Request: http.Request{
    Path: "/set",
    Headers: map[string]string{
      "Some-Other-Header": "val",
      "X-Header-Set":      "some-other-value",
    },
  },
  Headers: map[string]string{
    "Some-Other-Header": "val",
    "X-Header-Set":      "set-overwrites-values",
  },
  Backend:   "infra-backend-v1",
  Namespace: ns,
}
```

` "Some-Other-Header": "val",` is asserted to be a response header, but it shouldn't be. The test seems to conflate request headers and response headers

### PR DESCRIPTION
**Pull request by sunjayBhatia**
**What type of PR is this?**
/kind bug
/kind cleanup

**What this PR does / why we need it**:
Fixes logic in `HTTPRouteResponseHeaderModifier` conformance test.

Previously, the test conflated request and response headers.
Changes in this PR along with updated echoserver (see https://github.com/kubernetes-sigs/ingress-controller-conformance/pull/100) allow test to modify what response headers the backend app returns, to ensure modifications an implementation makes as a result of the ResponseHeaderModifier Filter are correct.

Also does some minor cleanup and changes/fixes assertions around headers.

This PR is verified with a local build of the echoserver image from https://github.com/kubernetes-sigs/ingress-controller-conformance/pull/100 as well as this spike on Contour to implement the ResponseHeaderModifier Filter: https://github.com/sunjayBhatia/contour/tree/gwapi-latest-with-resp-header-modification

Easiest to review this PR per-commit:
- first addresses #1384 to move expected response fields into their own struct
- second is the bulk of the change to fix the ResponseHeaderModifier test
- next is a little cleanup and update to changes in echoserver usage
- last is a change to the echoserver image, which will need to be updated once https://github.com/kubernetes-sigs/ingress-controller-conformance/pull/100 is merged and new echoserver image built

**Which issue(s) this PR fixes**:
Fixes #1473
Fixes #1384

**Does this PR introduce a user-facing change?**:
```release-note
NONE
```

### CODE REVIEW

**Comment by sunjayBhatia** [comment]
force pushed to fix commit message

**Comment by sunjayBhatia** [comment]
and again to rebase on main

File: conformance/base/manifests.yaml
Comment (howardjohn) [reply conformance/base/manifests.yaml:242]: I think we need to do that before merging...? otherwise this will break downstreams

Comment (sunjayBhatia) [reply conformance/base/manifests.yaml:242]: oh yeah for sure, the other PR for echoserver hasn't been merged though so I can't update this to anything valid yet until that is done https://github.com/kubernetes-sigs/ingress-controller-conformance/pull/100

File: conformance/utils/http/http.go
Comment (sunjayBhatia) [inline conformance/utils/http/http.go:43]: an alternative is to just have the tests specify the `X-Echo-Set-Header` header themselves, but this seemed like a better solution so people adding tests don't have to know how that exactly works

**Comment by sunjayBhatia** [comment]
/unhold

echoserver PR merged and image pushed to gcr

File: conformance/base/manifests.yaml
Comment (sunjayBhatia) [reply conformance/base/manifests.yaml]: echoserver reference updated to latest, resolving

**Review by howardjohn — APPROVED** [review]

**Comment by robscott** [comment]
Thanks @sunjayBhatia! This LGTM but would appreciate another set of eyes to confirm.

/cc @shaneutt @youngnick 
/approve

**Review by Xunzhuo — APPROVED** [review]

**Comment by howardjohn** [comment]
Tested on Istio and it passes now :+1: thanks!

**Pull request merged**
